hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala

```diff
@@ -651,6 +651,27 @@ extends Importer:
       Term.Error
     case Error() =>
       Term.Error
+    // A use[T] construct: analogous to Scala's summon[T].
+    case TermDef(Ins, Tree.Tup(tyArgs), N) => 
+      if tyArgs.length != 1 then
+        raise(ErrorReport(msg"Illegal use[T] construct. Only one type is allowed." -> tree.toLoc :: Nil))
+        Term.Error
+      val ty = term(tyArgs.head)
+      Term.Summon(ty)(tree, N)
+    // A funny parsing problem: 
+    //  - use[T].foo is parsed as (use ([T].foo) )
+    //  - use[T](...) is parsed as (use ([T]()) )
+    case TermDef(Ins, head, N) =>
+      // We try to pull out the [T] part from the tree.
+      def go(t: Tree): Tree = t match
+        case App(lhs, rhs) => App(go(lhs), rhs)
+        case InfixApp(lhs, kw, rhs) => InfixApp(go(lhs), kw, rhs)
+        case Sel(lhs, rhs) => Sel(go(lhs), rhs)
+        case SynthSel(lhs, rhs) => SynthSel(go(lhs), rhs)
+        case Jux(lhs, rhs) => Jux(go(lhs), rhs)
+        case tyArgs: Tup =>
+          TermDef(Ins, tyArgs, N)
+      term(go(head))
```

Comment:
Here is a funny fix of a parsing issue: `use[T].foo` actually gets parsed into `use ([T].foo)`, i.e.,

```fs
:pt
use[T].foo
//│ Parsed tree:
//│ TermDef:
//│   k = Ins
//│   head = Sel:
//│     prefix = Tup of Ls of 
//│       Ident of "T"
//│     name = Ident of "foo"
//│   rhs = N
```

This seems not unable to fix at the ParseRule level because `[T].foo` together forms an expression. The funny 🤡 workaround here is to recursively go down to the LHS of the trees, and "pull out" the `[T]` part from the tree.

I think this can somehow be fixed by adding a special case in the parser that only parses an expression surrounded by a pair of square brackets, but I think I wouldn't have time to study into how the parser works because of the _useless_ final exams 🫡

Reply:
oh no

Reply:
Following our discussion, let's change the syntax of used instances to `using T = t`.